CharacterizeTransformMatrix/CharacterizeTransformMatrix.py

```diff
@@ -0,0 +1,425 @@
+                "scipyRotationObject",
+                "rotationAxis",
+                "rotationAngleDegrees",
+                "eulerAnglesRAS",
+                "translationVector",
+                "translationOnlyMatrix",
+                "rotationOnlyMatrix",
+                "stretchOnlyMatrix",
+                "scaleMatrixList",
+                "stretchEigenvectorMatrix",
+            ],
+        )
+        results = resultsNamedTupleClass(
+            textResults=textResults,
+            isRigid=isRigid,
+            scaleFactors=f,
+            scaleDirections=scaleDirections,
+            largestPercentChangeScale=largestPercentChange,
+            volumePercentChangeOverall=volumePercentChange,
+            scipyRotationObject=r,
+            rotationAxis=rotation_axis,
+            rotationAngleDegrees=rotation_angle_deg,
+            eulerAnglesRAS=euler_angles_xyz,
+            translationVector=translationVector,
+            translationOnlyMatrix=T,
+            rotationOnlyMatrix=R,
+            stretchOnlyMatrix=K,
+            scaleMatrixList=S,
+            stretchEigenvectorMatrix=X,
+        )
+        return results
+
+    def separateTranslation(self, H):
+        """Given 4x4 numpy matrix H, decompose into a translation-only
+        matrix T and a no-translation matrix L, such that the matrix
+        product T*L is H.
+        """
+        T = np.eye(4)
+        T[:3, 3] = H[:3, 3]
+        L = H.copy()
+        L[:3, 3] = 0
+        assert np.allclose(H, T @ L), "T*L should equal H, but it does not!"
+        return T, L
+
+    def polarDecompose(self, H):
+        """Compute polar decomposition of 4x4 numpy matrix H.
+        Outputs are:
+        T: translation only matrix
+        R: rotation only matrix
+        K: stretch matrix
+        S: list of scale matrices, in order of decreasing eigenvalues
+        f: list of eigenvalues of the stretch matrix
+        X: eigenvector matrix of stretch matrix
+        The decomposition is such that the following is true:
+        T*R*K = H
+        S1*S2*S3 = K (where S<N> is the Nth scale matrix)
+        Furthermore, S1 is a matrix which scales in the direction of
+        the first eigenvector of K by a factor of the first (largest) eigenvalue
+        of K. Similarly, S2 is a matrix which scales in the direction of
+        the eigenvector associated with the second largest eigenvalue of K by
+        a factor of the second largest eigenvalue. Analgously for S3.
+        """
+        from scipy.linalg import polar
+
+        T, L = self.separateTranslation(H)  # T is translation matrix,
+        R, K = polar(L)  # R is rotation matrix, K is stretch matrix
+        # The determinant of a rotation matrix must be positive
+        if np.linalg.det(R) < 0:
+            R[:3, :3] = -R[:3, :3]
+            K[:3, :3] = -K[:3, :3]
+        # Check answer still OK
+        assert np.allclose(L, R @ K), "R*K should equal L, but it does not!"
```

Comment:
`assert` is intended for testing and debugging. Please replace all `assert`s in non-test code by checking of the condition and raising an exception if the check fails.

Reply:
Changed asserts to exceptions